server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpgradeService.java

```diff
@@ -76,90 +70,128 @@ public class MetadataUpgradeService {
     private final MetadataIndexUpgrader indexUpgrader;
     private final DocTableInfoFactory tableFactory;
     private final UserDefinedFunctionService userDefinedFunctionService;
-    private final IndexTemplateUpgrader templateUpgrader;
 
     public MetadataUpgradeService(NodeContext nodeContext,
                                   IndexScopedSettings indexScopedSettings,
                                   UserDefinedFunctionService userDefinedFunctionService) {
         this.tableFactory = new DocTableInfoFactory(nodeContext);
         this.indexScopedSettings = indexScopedSettings;
         this.indexUpgrader = new MetadataIndexUpgrader();
-        this.templateUpgrader = new IndexTemplateUpgrader();
         this.userDefinedFunctionService = userDefinedFunctionService;
     }
 
-    /**
-     * Elasticsearch 2.0 removed several deprecated features and as well as support for Lucene 3.x. This method calls
-     * {@link MetadataUpgradeService} to makes sure that indices are compatible with the current version. The
-     * MetadataIndexUpgradeService might also update obsolete settings if needed.
-     *
-     * @return input <code>metadata</code> if no upgrade is needed or an upgraded metadata
-     */
     public Metadata upgradeMetadata(Metadata metadata) {
-        boolean changed = false;
-        final Metadata.Builder upgradedMetadata = Metadata.builder(metadata);
-
-        // carries upgraded IndexTemplateMetadata to MetadataIndexUpgradeService.upgradeIndexMetadata
-        final Map<String, IndexTemplateMetadata> upgradedIndexTemplateMetadata = new HashMap<>();
-
-        // upgrade current templates
-        if (applyUpgrader(
-            metadata.templates(),
-            this::upgradeTemplates,
-            upgradedMetadata::removeTemplate,
-            (s, indexTemplateMetadata) -> {
-                upgradedIndexTemplateMetadata.put(s, indexTemplateMetadata);
-                upgradedMetadata.put(indexTemplateMetadata);
-            })) {
-            changed = true;
+        final Metadata.Builder newMetadata = Metadata.builder(metadata);
+
+        // Templates only exist in Metadata from < 6.1.0
+        // If streaming Metadata to nodes < 6.1.0 templates are re-created on demand
+        for (var cursor : metadata.templates()) {
+            String templateName = cursor.key;
+            newMetadata.removeTemplate(templateName);
+            if (templateName == IndexTemplateUpgrader.CRATE_DEFAULTS) {
+                continue;
+            }
+            IndexTemplateMetadata template = cursor.value;
+            RelationName relationName = IndexName.decode(template.name()).toRelationName();
+            RelationMetadata relation = metadata.getRelation(relationName);
+            assert relation == null
+                : "If there is still a template present there shouldn't be any RelationMetadata";
```

Comment:
I'm not sure about this. Before it used the `relation.tableVersion()` in favour of the `docTable.tableVersion()`. But no test is triggering this assertion.
Afaik it also shouldn't happen during rolling upgrades?

Reply:
Just to add: I think it can never happen the the metadata contains both, templates *and* relations. If the master still operates on templates, the received state will do as well and will not contain any relation. Such the assertion would never be triggered like you observed.